source/grid/grid_tools.cc

```diff
@@ -2125,6 +2191,155 @@ namespace GridTools
+                                                   current_proc_idx,
+                                                   current_cell_idx,
+                                                   n_active_cellls,
+                                                   n_partitions);
+        }
+    }
+  }
+
+  template <int dim, int spacedim>
+  void
+  partition_triangulation_zorder (const unsigned int          n_partitions,
+                                  Triangulation<dim,spacedim> &triangulation)
+  {
+    Assert ((dynamic_cast<parallel::distributed::Triangulation<dim,spacedim>*>
+             (&triangulation)
+             == 0),
+            ExcMessage ("Objects of type parallel::distributed::Triangulation "
+                        "are already partitioned implicitly and can not be "
+                        "partitioned again explicitly."));
+    Assert (n_partitions > 0, ExcInvalidNumberOfPartitions(n_partitions));
+
+    // check for an easy return
+    if (n_partitions == 1)
+      {
+        for (typename dealii::internal::ActiveCellIterator<dim, spacedim, Triangulation<dim, spacedim> >::type
+             cell = triangulation.begin_active();
+             cell != triangulation.end(); ++cell)
+          cell->set_subdomain_id (0);
+        return;
+      }
+
+    typedef typename Triangulation<dim,spacedim>::active_cell_iterator ac_it;
+    typedef typename Triangulation<dim,spacedim>::cell_iterator lvl_it;
+
+    unsigned int current_proc_idx=0;
+    unsigned int current_cell_idx=0;
+    const unsigned int n_active_cellls = triangulation.n_active_cells();
+
+    // create coarse cell reordering
+    std::vector<unsigned int> coarse_cell_to_p4est_tree_permutation;
+    std::vector<unsigned int> p4est_tree_to_coarse_cell_permutation;
+
+    DynamicSparsityPattern cell_connectivity;
+    GridTools::get_vertex_connectivity_of_cells_on_level (triangulation, 0, cell_connectivity);
+    coarse_cell_to_p4est_tree_permutation.resize (triangulation.n_cells(0));
+    SparsityTools::reorder_hierarchical (cell_connectivity,
+                                         coarse_cell_to_p4est_tree_permutation);
+
+    p4est_tree_to_coarse_cell_permutation
+      = Utilities::invert_permutation (coarse_cell_to_p4est_tree_permutation);
+
+    // set subdomain id for active cell descendants
+    // of each coarse cell in permuted order
+    for (unsigned int idx=0; idx<triangulation.n_cells(0); ++idx)
+      {
+        const unsigned int coarse_cell_idx = p4est_tree_to_coarse_cell_permutation[idx];
+        lvl_it coarse_cell (&triangulation, 0, coarse_cell_idx);
+        set_subdomain_id_in_zorder_recursively(coarse_cell,
+                                               current_proc_idx,
+                                               current_cell_idx,
+                                               n_active_cellls,
+                                               n_partitions);
+      }
+
+    // ensure terminal siblings are on the same
```

Comment:
can you expand this to say:
If all children of a cell are active (so we have a cell that is refined once and no part is refined further), p4est places all of them on the same processor. The new owner will be the processor with the largest number of children (ties are broken by picking the lower rank). Duplicate this logic here.